corehq/motech/openmrs/atom_feed.py

```diff
@@ -0,0 +1,269 @@
+from corehq.motech.openmrs.repeaters import OpenmrsRepeater
+from corehq.motech.requests import Requests
+from corehq.util.soft_assert import soft_assert
+
+
+UuidUpdatedAt = namedtuple('UuidUpdatedAt', 'uuid updated_at')
+_assert = soft_assert(['@'.join(('nhooper', 'dimagi.com'))])
+
+
+def get_patient_feed_xml(requests, page):
+    if not page:
+        # If this is the first time the patient feed is polled, just get
+        # the most recent changes. This shows updating patients
+        # successfully, but does not replay all OpenMRS changes.
+        page = 'recent'
+    patient_feed_url = '/ws/atomfeed/patient/' + page
+    resp = requests.get(patient_feed_url)
+    root = etree.fromstring(resp.content)
+    return root
+
+
+def get_timestamp(element, xpath='./atom:updated'):
+    """
+    Returns a datetime instance of the text at the given xpath.
+
+    >>> element = etree.XML('''<feed xmlns="http://www.w3.org/2005/Atom">
+    ...     <updated>2018-05-15T14:02:08Z</updated>
+    ... </feed>''')
+    >>> get_timestamp(element)
+    datetime.datetime(2018, 5, 15, 14, 2, 8, tzinfo=tzutc())
+
+    """
+    timestamp_elems = element.xpath(xpath, namespaces={'atom': 'http://www.w3.org/2005/Atom'})
+    if not timestamp_elems:
+        raise ValueError('XPath "{}" not found'.format(xpath))
+    if len(timestamp_elems) != 1:
+        raise ValueError('XPath "{}" matched multiple nodes'.format(xpath))
+    tzinfos = {'UTC': tzutc()}
+    return dateutil_parser.parse(timestamp_elems[0].text, tzinfos=tzinfos)
+
+
+def get_patient_uuid(element):
+    """
+    Extracts the UUID of a patient from an entry's "content" node.
+
+    >>> element = etree.XML('''<entry>
+    ...     <content type="application/vnd.atomfeed+xml">
+    ...         <![CDATA[/openmrs/ws/rest/v1/patient/e8aa08f6-86cd-42f9-8924-1b3ea021aeb4?v=full]]>
+    ...     </content>
+    ... </entry>''')
+    >>> get_patient_uuid(element) == 'e8aa08f6-86cd-42f9-8924-1b3ea021aeb4'
+    True
+
+    """
+    # "./*[local-name()='content']" ignores namespaces and matches all
+    # child nodes with tag name "content". This lets us traverse the
+    # feed regardless of whether the Atom namespace is explicitly given.
+    content = element.xpath("./*[local-name()='content']")
+    pattern = re.compile(r'/patient/([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})\b')
+    if content and len(content) == 1:
+        cdata = content[0].text
+        matches = pattern.search(cdata)
+        if matches:
+            return matches.group(1)
+    raise ValueError('patient UUID not found')
+
+
+def get_updated_patients(repeater):
+    """
+    Iterates over a paginated atom feed, yields patients updated since
+    repeater.atom_feed_last_polled_at, and updates the repeater.
+    """
+    requests = Requests(
+        repeater.domain,
+        repeater.url,
+        repeater.username,
+        repeater.password
+    )
+    last_polled_at = pytz.utc.localize(repeater.atom_feed_last_polled_at)
```

Comment:
why is this localization necessary?

Reply:
The Atom Feed's timestamps are timezone-aware. So when we compare timestamps (in lines 109, 117) this timestamp also needs to be timezone-aware. `atom_feed_last_polled_at` is a UTC timestamp (`datetime.utcnow()`) but the timezone gets dropped by `DateTimeProperty`.